From 1d5b1cf4278a2a4438797085226bd2575caad12a Mon Sep 17 00:00:00 2001
From: Jeremy Kerr <jk@codeconstruct.com.au>
Date: Fri, 18 Feb 2022 12:25:54 +0800
Subject: [PATCH backport 5.10 56/73] mctp: add address validity checking for
 packet receive

This change adds some basic sanity checks for the source and dest
headers of packets on initial receive.

Signed-off-by: Jeremy Kerr <jk@codeconstruct.com.au>
Signed-off-by: Jakub Kicinski <kuba@kernel.org>
(cherry picked from commit 86cdfd63f25dc1c8f241ee70c58da3c10472b76e)
---
 net/mctp/route.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/net/mctp/route.c b/net/mctp/route.c
index 673fa21645f9..94d41b45058c 100644
--- a/net/mctp/route.c
+++ b/net/mctp/route.c
@@ -1091,6 +1091,17 @@ static int mctp_pkttype_receive(struct sk_buff *skb, struct net_device *dev,
 	if (mh->ver < MCTP_VER_MIN || mh->ver > MCTP_VER_MAX)
 		goto err_drop;
 
+	/* source must be valid unicast or null; drop reserved ranges and
+	 * broadcast
+	 */
+	if (!(mctp_address_unicast(mh->src) || mctp_address_null(mh->src)))
+		goto err_drop;
+
+	/* dest address: as above, but allow broadcast */
+	if (!(mctp_address_unicast(mh->dest) || mctp_address_null(mh->dest) ||
+	      mctp_address_broadcast(mh->dest)))
+		goto err_drop;
+
 	/* MCTP drivers must populate halen/haddr */
 	if (dev->type == ARPHRD_MCTP) {
 		cb = mctp_cb(skb);
-- 
2.20.1

